### ISSUE TITLE
Questions on the first stage of ``make``

### ISSUE THREAD

**Issue by kendonB** [wlandau-lilly/drake#121]
Hi again! 

I have a large project with ~30000 targets. When I run ``make``, it takes several minutes for the first piece of visual feedback to print (I see packages loading after these first few minutes). 

drake is doing something in these minutes. 

1) Out of curiosity, what is it doing? 
2) is there a visual feedback like a progress bar that might make sense for this first stage? 
3) Are there any potential gains in efficiency you could think about that wouldn't compromise robustness?

**Comment by kendonB** [comment]
Did some digging and it seems to be the line:

``command_deps <- lapply(plan$command, command_dependencies)`` in ``build_graph()``

**Comment by kendonB** [comment]
Looking through what the function actually does, it's not immediately clear what you can do to speed this up, other than porting the function and subfunctions to C++.

**Comment by wlandau-lilly** [comment]
@kendonB I was hoping for a huge test case like this!

You may be happy to know that [that line of code](https://github.com/wlandau-lilly/drake/blob/master/R/graph.R#L47) now reads

```r
command_deps <- lightly_parallelize(
    plan$command, command_dependencies, jobs = jobs)
```

Since you're not using a Windows machine, you can just set `jobs` to be whatever you want `mc.cores` to be so that `mclapply()` speeds things up. Since you're using the `"future_lapply"` backend for targets, this will not affect how the meat of your workflow is processed.

To speed things up further, I think we would need a C++-accelerated static code analysis tool to parse commands and look for dependencies (targets/imports). `Drake` uses `codetools` plus some customizations, and the state of the art is `CodeDepends`. Neither has compiled code.

Do you think `drake` should log what it's doing to the console? It might be too much to print one line per target, but I was thinking something like `resolve dependencies of imports` and `resolve dependencies of targets`.

You should also check out the [new sub-graphing functionality](https://github.com/wlandau-lilly/drake/issues/88#issuecomment-340148870) I implemented over the weekend. As long as you don't have too many imports, the [interactive visNetwork graph](http://datastorm-open.github.io/visNetwork/) should be sane, even with a 30000-target project. You might even speed things up by using the `targets` argument to `plot_graph()`.

**Comment by wlandau-lilly** [comment]
[Your comment here](https://github.com/wlandau-lilly/drake/issues/121#issuecomment-340646861) is an interesting thought. Currently, `drake` does not try to pull out unique code fragments or unique imports to analyze. Intuitively, I don't think having unique values would noticeably increase speed, and `unique` could take a long time on 30000 targets. On the other hand, I do not have projects as large as yours, so feel free to prove me wrong.

**Comment by kendonB** [comment]
I think that printing a line like ``resolve dependencies of imports`` or ``resolve dependencies of targets`` makes sense.

Regarding pulling out unique functions, I imagine most large projects are, like mine, just many small chunks of few operations. For example, I have 6 user-written functions generating the 30000 targets. On mine, it would greatly increase speed!

**Comment by wlandau-lilly** [comment]
I think I see. In a workflow plan with 30000 targets

```
##    target       command
## 1  target1  rnorm(1000)
## 2  target2  rnorm(1000)
...
```

"rnorm(1000)" should really only be processed once.

**Comment by kendonB** [comment]
You could go finer than that though. Even with commands ``my_fun(obj1)``, and ``my_fun(obj2)`` you could check the body of ``my_fun`` once.

**Comment by wlandau-lilly** [comment]
Actually, `drake` already does that last part just fine. It doesn't waste any time processing duplicate imports. Each code analysis only goes one level deep at a time (the body of one function or the code fragment of one command).

**Comment by kendonB** [comment]
You could also consider caching results so that future runs of make don't do the same work twice? Kind of like ``memoise`` but using the disk?

**Comment by kendonB** [comment]
Another potential speed up for usability could be to implement different levels of checking like ``remake``? This is really nice while testing.

**Comment by wlandau-lilly** [comment]
`Drake` does already cache imports, but it always checks for changes. It always needs to check for changes somehow. I would assume `storr` does not cache items if their hashes have not changed, but that is a question for @richfitz.

What levels of checking would you like to see? If you have custom file targets/imports like `report.md` and `report.Rmd` in the basic example, `drake` uses file modification times to check whether to bother computing a new hash (#4), but that's about it.

**Comment by kendonB** [comment]
I quite liked remakes different levels: "exists" (current if the target exists), "depends" (current if exists and dependencies unchanged), "code" (current if exists and code unchanged) or "all" (current if exists and both dependencies and code unchanged).

Regarding caching, I was referring to the output of ``command_dependencies`` which seems different from caching imports?

**Comment by wlandau-lilly** [comment]
I did not know `remake` exposed those levels of checking to the user, and I still do not think I see the full practical use. For `drake`, dependency checking is mostly in [current.R](https://github.com/wlandau-lilly/drake/blob/master/R/current.R) and [hash.R](https://github.com/wlandau-lilly/drake/blob/master/R/hash.R).

I was not actually sure about caching all the individual command dependencies for every target. I opted for a small manageable [dependency hash](https://github.com/wlandau-lilly/drake/blob/master/R/hash.R#L19), thinking it would cut down on storage and time spent caching.

**Comment by wlandau-lilly** [comment]
By the way, b18b43c89f91a3b46b37e39d921f0709d62424c1 and beyond has those console messages. Now, more people will be able to guess why [build_graph()](https://github.com/wlandau-lilly/drake/blob/master/R/graph.R#L26) takes so long.

```r
> load_basic_example()
> make(my_plan)
interconnect 7 items: ld, td, simulate, reg1, my_plan, reg2, dl
interconnect 15 items: 'report.md', small, large, regression1_small, regressi...
import knit
import 'report.Rmd'
...
```

**Comment by wlandau-lilly** [comment]
I could probably [memoise](https://github.com/r-lib/memoise) the output of `command_dependencies()`. The output of `import_dependencies()` could be quite large and not worth the trouble.

**Comment by wlandau-lilly** [comment]
The memoization is straightforward. Environments might give us uniqueness since they're designed to avoid deep copies.

**Comment by wlandau-lilly** [comment]
See [this new Stack Overflow post](https://stackoverflow.com/questions/47037428/versions-of-lapply-and-mclapply-that-avoid-redundant-processing). This seems like such a common problem, and I hope someone has solved it already.

**Comment by wlandau-lilly** [comment]
In , see [lightly_parallelize()](https://github.com/wlandau-lilly/drake/blob/master/R/parallel.R#L41), which now calls [lightly_parallelize_atomic()](https://github.com/wlandau-lilly/drake/blob/master/R/parallel.R#L50). Should avoid duplication of effort in the [processing of commands](https://github.com/wlandau-lilly/drake/blob/master/R/graph.R#L55). I posted the [general solution here](https://stackoverflow.com/a/47038331/3704549).

**Comment by wlandau-lilly** [comment]
I think the last thing on this thread is the possible memoization of [processing commands here](https://github.com/wlandau-lilly/drake/blob/master/R/graph.R#L55). [memoise](https://github.com/r-lib/memoise) with the file system cache is probably the best option. I am a bit reluctant because this memoization cache belongs inside the `.drake/` `storr` cache, and passing the right path to [build_graph()](https://github.com/wlandau-lilly/drake/blob/master/R/graph.R#L26) would be a pain.

Also, the code to memoise is this:

```r
command_deps <- lightly_parallelize(
    plan$command, command_dependencies, jobs = jobs)
```

It depends on both the commands and the number of jobs. I would want to memoize in a way that ignores `jobs`.

**Comment by wlandau-lilly** [comment]
See r-lib/memoise#54. There is resistance to optionally ignoring parameters like `jobs`.

**Comment by wlandau-lilly** [comment]
I will memoize only if the `memoise` package handles [this use case](https://github.com/r-lib/memoise/issues/54#issuecomment-340802443) reasonably well.

**Comment by wlandau-lilly** [comment]
I think I have done all I can do on this thread at the moment. Will reopen if/when there are new possibilities or suggestions.

**Comment by kendonB** [comment]
Ran this again and it runs much faster with the parallelization. I noticed that the ``interconnect`` message came up twice for the same set of targets. Is this what you expect? It appears first then I see a bunch of ``import ...``, then I see ``interconnect`` again, then ``check ...``, then the making happens.

**Comment by wlandau-lilly** [comment]
Glad to hear the parallelism was an improvement. I expect your idea of [lightly_parallelize_atomic()](https://github.com/wlandau-lilly/drake/blob/master/R/parallel.R#L41) may have sped things up too (avoid analyzing the same workflow command twice). I do expect `interconnect` to show up twice: one for imports only, and another for the targets listed in your workflow plan. There are two calls to `console_dependencies()`, one [here](https://github.com/wlandau-lilly/drake/blob/master/R/graph.R#L45) and another [here](https://github.com/wlandau-lilly/drake/blob/master/R/graph.R#L51). I suppose I should make different console messages for each. Do you see two `interconnect`s for the same `make()`?

**Comment by wlandau-lilly** [comment]
FYI: I just made the distinction between the two `interconnect` (now `connect`) logs clearer. Now, one says `imports` and the other says `targets`.

```r
load_basic_example()
make(my_plan)
```

```
## cache /lrlhps/users/c240390/.drake
## connect 9 imports: ld, lint_summary, gwd, td, simulate, reg1, my_plan, reg2, dl
## connect 15 targets: 'report.md', small, large, regression1_small, regression1...
## check 9 items: 'report.Rmd', knit, summary, suppressWarnings, coefficients, d...
## import 'report.Rmd'
## import knit
## import summary
...
```

Do you still get `import ...` messages before the `connect` ones?

**Comment by kendonB** [comment]
I actually see more than two (names removed):

```
cache ....
interconnect 42 items: ...
interconnect 24306 items: ...
interconnect 42 items: ...
interconnect 24306 items: ...
check 92 items: ...
import ... lots
interconnect 42 items: ...
interconnect 24306 items: ...
check 92 items: ...
check 7 items: ...
```

**Comment by wlandau-lilly** [comment]
I understand.

I actually reproduced this on one of my larger projects. Happens with distributed parallelism only. Will explain tonight or tomorrow.

**Comment by kendonB** [comment]
I also see the same behavior when running ``drake::outdated(my_plan, jobs = 8)`` so it seems to also happen with vanilla parallelism?

**Comment by wlandau-lilly** [comment]
Yes, I am seeing redundant work in `outdated()` through `config()`. Currently working on refactoring the internals.

**Comment by wlandau-lilly** [comment]
@kendonB you just rooted out major inefficiencies in `drake`. The fixes are effective as of 29f6d29badd92fc3830f0583c5f274923df591b4. Not only did those extra `connect` messages go away, but the unit tests became blazing fast! This makes my life so much easier, and I cannot thank you enough! If it sounds good to you, please submit a pull request to add yourself as a contributor above [this line in the DESCRIPTION](https://github.com/wlandau-lilly/drake/blob/master/DESCRIPTION#L15).

**Comment by kendonB** [comment]
No need to add me as a contributor. Thanks for the fixes.
